### PR TITLE
[Dataframes] Implemented .describe()

### PR DESCRIPTION
**Pull request by 11rohans**
Implemented methods necessary for describe. Currently uses the transpose trick, and can be optimized further.  
mean() 
std() 
var() 
median() 
quantile() 
describe()

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 import pandas as pd
 from pandas.api.types import is_scalar
 from pandas.util._validators import validate_bool_kwarg
-from pandas.core.index import _ensure_index_from_sequences
+# from pandas.core.index import _ensure_index_from_sequences
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: why did this get commented out? If we're no longer using it then let's just remove it.

File: python/ray/dataframe/test/test_dataframe.py [hunk 2]
```diff
@@ -1836,11 +1862,12 @@ def test_transform():
         ray_df.transform(None)
 
 
+@pytest.fixture
 def test_truediv():
     ray_df = create_test_dataframe()
 
     with pytest.raises(NotImplementedError):
-        ray_df.truediv(None)
+        ray_df.rtruediv(None)
```
Comment (devin-petersohn) [inline python/ray/dataframe/test/test_dataframe.py]: `ray_df.rtruediv(None)` -> `ray_df.truediv(None)`

File: python/ray/dataframe/test/test_dataframe.py [hunk 3]
```diff
@@ -1836,11 +1862,12 @@ def test_transform():
         ray_df.transform(None)
 
 
+@pytest.fixture
```
Comment (devin-petersohn) [inline python/ray/dataframe/test/test_dataframe.py]: Please remove this line also.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -705,9 +705,42 @@ def cumsum(self, axis=None, skipna=True, *args, **kwargs):
             "github.com/ray-project/ray.")
 
     def describe(self, percentiles=None, include=None, exclude=None):
-        raise NotImplementedError(
-            "To contribute to Pandas on Ray, please visit "
-            "github.com/ray-project/ray.")
+        """
+        Generates descriptive statistics that summarize the central tendency,
+        dispersion and shape of a dataset’s distribution, excluding NaN values.
+
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please also add docs for the `Args` and `Return`.

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -1405,9 +1466,27 @@ def product(self, axis=None, skipna=None, level=None, numeric_only=None,
 
     def quantile(self, q=0.5, axis=0, numeric_only=True,
                  interpolation='linear'):
-        raise NotImplementedError(
-            "To contribute to Pandas on Ray, please visit "
-            "github.com/ray-project/ray.")
+        """Return values at the given quantile over requested axis,
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please also add description of args.

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
Looks great! Just a couple minor formatting things, and you forgot doc components on a couple.

File: python/ray/dataframe/dataframe.py [hunk 6]
```diff
@@ -718,9 +718,80 @@ def cumsum(self, axis=None, skipna=True, *args, **kwargs):
...
+        the result will include a union of attributes of each type.
+
+        The include and exclude parameters can be used to limit which columns
+        in a DataFrame are analyzed for the output.
+        The parameters are ignored when analyzing a Series.
+
+        Returns: Series/DataFrame of summary statistics
+        """
+
+        obj_columns = []
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: prefer to change to `obj_columns = [self.columns[i] for i, t in enumerate(self.dtypes) if t == np.dtype('O')]`

File: python/ray/dataframe/dataframe.py (hunk 4 above)
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: Please format the comment as follows:
```
Args:
    percentiles: ...
    ...

Returns:
    ...
```

File: python/ray/dataframe/dataframe.py [hunk 7]
```diff
@@ -1513,15 +1584,65 @@ def max(self, axis=None, skipna=None, level=None, numeric_only=None,
...
-            "To contribute to Pandas on Ray, please visit "
-            "github.com/ray-project/ray.")
+        """Computes mean across the DataFrame.
+
+        Args:
+            axis (int): The axis to take the mean on.
+            skipna (bool): True to skip NA values, false otherwise.
+
+        Returns:
+<<<<<<< HEAD
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please remove merge artifact.

File: python/ray/dataframe/dataframe.py [hunk 8]
```diff
@@ -1513,15 +1584,65 @@ def max(self, axis=None, skipna=None, level=None, numeric_only=None,
...
+                                axis=1, skipna=skipna,
+                                level=level, numeric_only=numeric_only
+                              )
+        else:
+            func = (lambda df: df.T.mean(axis=0,
+                    skipna=None, level=None, numeric_only=None))
+
+            new_df = [_deploy_func.remote(func, part) for part in self._df]
+
+            items = [ray.get(item) for item in new_df]
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: You can just call `ray.get(new_df)`.

File: python/ray/dataframe/dataframe.py [hunk 9]
```diff
@@ -1513,15 +1584,65 @@ def max(self, axis=None, skipna=None, level=None, numeric_only=None,
...
+        if axis == 0 or axis is None:
+            return self.T.mean(
+                                axis=1, skipna=skipna,
+                                level=level, numeric_only=numeric_only
+                              )
+        else:
+            func = (lambda df: df.T.mean(axis=0,
+                    skipna=None, level=None, numeric_only=None))
+
+            new_df = [_deploy_func.remote(func, part) for part in self._df]
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Prefer `new_df` change to `computed_means`.

File: python/ray/dataframe/dataframe.py [hunk 10]
```diff
@@ -1513,15 +1584,65 @@ def max(self, axis=None, skipna=None, level=None, numeric_only=None,
...
+
+            items = [ray.get(item) for item in new_df]
+            r1 = items[0]
+            r_other = [items[i] for i in range(1, len(items))]
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: What are you doing here? Is this a flatten?

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
I have a few questions on the implementation that I missed the first time.

File: python/ray/dataframe/dataframe.py (hunk 5 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: resolved

File: python/ray/dataframe/dataframe.py (hunk 4 above)
Comment (11rohans) [reply python/ray/dataframe/dataframe.py]: resolved

**Review by devin-petersohn — APPROVED** [review]

**Pull request merged**